File: apps/backend/src/routes/jira.ts
import { Hono } from "hono";
import {
  jiraIssuesRequestSchema,
  jiraTestRequestSchema,
} from "../lib/validation";

const jira = new Hono();

const normalizeBaseUrl = (baseUrl: string) => {
  const url = new URL(baseUrl);
  return `${url.protocol}//${url.hostname}`;
};

const buildBasicAuth = (email: string, apiToken: string) => {
  const raw = `${email}:${apiToken}`;
  return Buffer.from(raw).toString("base64");
};

const extractJiraError = (data: unknown, status: number) => {
  if (data && typeof data === "object") {
    const typed = data as {
      errorMessages?: string[];
      errors?: Record<string, string>;
      message?: string;
    };
    if (typed.errorMessages?.length) return typed.errorMessages[0];
    if (typed.message) return typed.message;
    if (typed.errors) {
      const first = Object.values(typed.errors)[0];
      if (first) return first;
    }
  }
  return `Jira request failed: ${status}`;
};

const readJsonBody = async (c: { req: { json: () => Promise<unknown> } }) => {
  try {
    const body = await c.req.json();
    return { ok: true as const, body };
  } catch {
    return { ok: false as const };
  }
};

const requestJira = async (
  url: string,
  auth: string,
  options?: { method?: string; body?: string }
) => {
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), 8000);

  try {
    const response = await fetch(url, {
      method: options?.method,
      body: options?.body,
      headers: {
        Accept: "application/json",
        ...(options?.body ? { "Content-Type": "application/json" } : {}),
        Authorization: `Basic ${auth}`,
      },
      redirect: "error",
      signal: controller.signal,
    });
    const data = await response.json().catch(() => ({}));
    return { response, data };
  } finally {
    clearTimeout(timeoutId);
  }
};

/**
 * POST /integrations/jira/test
 * Test Jira Cloud API token credentials.
 */
jira.post("/integrations/jira/test", async (c) => {
  const parsedBody = await readJsonBody(c);
  if (!parsedBody.ok) {
    return c.json({ error: "Invalid JSON body" }, 400);
  }
  const parsed = jiraTestRequestSchema.safeParse(parsedBody.body);

  if (!parsed.success) {
    return c.json(
      { error: parsed.error.issues[0]?.message ?? "Invalid request" },
      400
    );
  }

  const { baseUrl, email, apiToken } = parsed.data;
  const url = `${normalizeBaseUrl(baseUrl)}/rest/api/3/myself`;
  const auth = buildBasicAuth(email, apiToken);

  try {
    const { response, data } = await requestJira(url, auth);

    if (!response.ok) {
      const status =
        response.status >= 400 && response.status < 500 ? 400 : 502;
      return c.json({ error: extractJiraError(data, response.status) }, status);
    }

    const profile = data as {
      displayName?: string;
      emailAddress?: string;
      accountId?: string;
      avatarUrls?: Record<string, string>;
    };

    return c.json({
      profile: {
        displayName: profile.displayName ?? "Unknown",
        emailAddress: profile.emailAddress,
        accountId: profile.accountId,
        avatarUrls: profile.avatarUrls,
      },
    });
  } catch (error) {
    const message =
      error instanceof Error && error.name === "AbortError"
        ? "Jira request timed out"
        : "Unable to reach Jira API";
    return c.json({ error: message }, 502);
  }
});

/**
 * POST /integrations/jira/issues
 * Fetch recent issues assigned to the current user.
 */
jira.post("/integrations/jira/issues", async (c) => {
  const parsedBody = await readJsonBody(c);
  if (!parsedBody.ok) {
    return c.json({ error: "Invalid JSON body" }, 400);
  }
  const parsed = jiraIssuesRequestSchema.safeParse(parsedBody.body);

  if (!parsed.success) {
    return c.json(
      { error: parsed.error.issues[0]?.message ?? "Invalid request" },
      400
    );
  }

  const { baseUrl, email, apiToken } = parsed.data;
  const url = `${normalizeBaseUrl(baseUrl)}/rest/api/3/search/jql`;
  const auth = buildBasicAuth(email, apiToken);
  const body = JSON.stringify({
    jql: "assignee = currentUser() ORDER BY updated DESC",
    maxResults: 20,
    fields: ["key", "summary", "status", "assignee", "updated"],
  });

  try {
    const { response, data } = await requestJira(url, auth, {
      method: "POST",
      body,
    });

    if (!response.ok) {
      const status =
        response.status >= 400 && response.status < 500 ? 400 : 502;
      return c.json({ error: extractJiraError(data, response.status) }, status);
    }

    const payload = data as {
      issues?: Array<{
        key?: string;
        fields?: {
          summary?: string;
          status?: { name?: string };
          updated?: string;
          assignee?: { displayName?: string };
        };
      }>;
    };

    const issues = (payload.issues ?? [])
      .map((issue) => ({
        key: issue.key ?? "",
        summary: issue.fields?.summary ?? "Untitled issue",
        status: issue.fields?.status?.name ?? "Unknown",
        updated: issue.fields?.updated,
        assignee: issue.fields?.assignee?.displayName,
      }))
      .filter((issue) => issue.key);

    return c.json({ issues });
  } catch (error) {
    const message =
      error instanceof Error && error.name === "AbortError"
        ? "Jira request timed out"
        : "Unable to reach Jira API";
    return c.json({ error: message }, 502);
  }
});

export default jira;

File: src-tauri/src/application/jira.rs
use base64::{engine::general_purpose, Engine as _};
use reqwest::blocking::Client;
use reqwest::redirect::Policy;
use reqwest::Method;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::time::Duration;
use url::Url;

#[derive(Debug, Deserialize, specta::Type)]
#[serde(rename_all = "camelCase")]
pub struct JiraRequestInput {
    pub base_url: String,
    pub email: String,
    pub api_token: String,
}

#[derive(Debug, Serialize, specta::Type)]
#[serde(rename_all = "camelCase")]
pub struct JiraUserProfile {
    pub display_name: String,
    pub email_address: Option<String>,
    pub account_id: Option<String>,
    pub avatar_urls: Option<std::collections::HashMap<String, String>>,
}

#[derive(Debug, Serialize, specta::Type)]
pub struct JiraTestConnectionResponse {
    pub profile: JiraUserProfile,
}

#[derive(Debug, Serialize, specta::Type)]
pub struct JiraIssue {
    pub key: String,
    pub summary: String,
    pub status: String,
    pub updated: Option<String>,
    pub assignee: Option<String>,
}

#[derive(Debug, Serialize, specta::Type)]
pub struct JiraIssuesListResponse {
    pub issues: Vec<JiraIssue>,
}

fn validate_email(email: &str) -> Result<(), String> {
    let trimmed = email.trim();
    if trimmed.is_empty() || !trimmed.contains('@') {
        return Err("Invalid email address".to_string());
    }
    Ok(())
}

fn validate_api_token(token: &str) -> Result<(), String> {
    if token.trim().is_empty() {
        return Err("API token is required".to_string());
    }
    Ok(())
}

fn normalize_base_url(base_url: &str) -> Result<String, String> {
    let parsed = Url::parse(base_url.trim()).map_err(|_| "Invalid Jira base URL".to_string())?;

    if parsed.scheme() != "https" {
        return Err("Jira base URL must start with https://".to_string());
    }

    let hostname = parsed
        .host_str()
        .ok_or_else(|| "Invalid Jira base URL".to_string())?;

    if !hostname.ends_with(".atlassian.net") {
        return Err(
            "Jira base URL must be a Jira Cloud site (e.g. https://your-domain.atlassian.net)"
                .to_string(),
        );
    }

    let path_ok = parsed.path().is_empty() || parsed.path() == "/";
    if !path_ok
        || !parsed.username().is_empty()
        || parsed.password().is_some()
        || parsed.port().is_some()
        || parsed.query().is_some()
        || parsed.fragment().is_some()
    {
        return Err(
            "Jira base URL must be a Jira Cloud site (e.g. https://your-domain.atlassian.net)"
                .to_string(),
        );
    }

    Ok(format!("{}://{}", parsed.scheme(), hostname))
}

fn build_basic_auth(email: &str, api_token: &str) -> String {
    general_purpose::STANDARD.encode(format!("{}:{}", email.trim(), api_token.trim()))
}

fn extract_jira_error(data: &Value, status: u16) -> String {
    if let Some(error_messages) = data.get("errorMessages").and_then(Value::as_array) {
        if let Some(first) = error_messages.first().and_then(Value::as_str) {
            if !first.trim().is_empty() {
                return first.to_string();
            }
        }
    }

    if let Some(message) = data.get("message").and_then(Value::as_str) {
        if !message.trim().is_empty() {
            return message.to_string();
        }
    }

    if let Some(errors) = data.get("errors").and_then(Value::as_object) {
        if let Some(first) = errors.values().find_map(Value::as_str) {
            if !first.trim().is_empty() {
                return first.to_string();
            }
        }
    }

    format!("Jira request failed: {}", status)
}

fn create_jira_client() -> Result<Client, String> {
    Client::builder()
        .timeout(Duration::from_secs(8))
        .redirect(Policy::none())
        .build()
        .map_err(|error| format!("Failed to create Jira client: {}", error))
}

fn request_jira(
    client: &Client,
    method: Method,
    url: &str,
    auth: &str,
    body: Option<Value>,
) -> Result<(u16, Value), String> {
    let mut request = client
        .request(method, url)
        .header("Accept", "application/json")
        .header("Authorization", format!("Basic {}", auth));

    if let Some(body) = body {
        request = request
            .header("Content-Type", "application/json")
            .json(&body);
    }

    let response = request.send().map_err(|error| {
        if error.is_timeout() {
            "Jira request timed out".to_string()
        } else {
            "Unable to reach Jira API".to_string()
        }
    })?;

    let status = response.status().as_u16();
    let data = response.json::<Value>().unwrap_or_else(|_| json!({}));
    Ok((status, data))
}

pub fn jira_test_connection(
    params: JiraRequestInput,
) -> Result<JiraTestConnectionResponse, String> {
    validate_email(&params.email)?;
    validate_api_token(&params.api_token)?;
    let normalized_base_url = normalize_base_url(&params.base_url)?;

    let client = create_jira_client()?;
    let auth = build_basic_auth(&params.email, &params.api_token);
    let url = format!("{}/rest/api/3/myself", normalized_base_url);
    let (status, data) = request_jira(&client, Method::GET, &url, &auth, None)?;

    if status >= 400 {
        return Err(extract_jira_error(&data, status));
    }

    let avatar_urls = data
        .get("avatarUrls")
        .and_then(Value::as_object)
        .map(|avatar_map| {
            avatar_map
                .iter()
                .filter_map(|(key, value)| value.as_str().map(|url| (key.clone(), url.to_string())))
                .collect::<std::collections::HashMap<_, _>>()
        });

    Ok(JiraTestConnectionResponse {
        profile: JiraUserProfile {
            display_name: data
                .get("displayName")
                .and_then(Value::as_str)
                .unwrap_or("Unknown")
                .to_string(),
            email_address: data
                .get("emailAddress")
                .and_then(Value::as_str)
                .map(ToOwned::to_owned),
            account_id: data
                .get("accountId")
                .and_then(Value::as_str)
                .map(ToOwned::to_owned),
            avatar_urls,
        },
    })
}

pub fn jira_list_issues(params: JiraRequestInput) -> Result<JiraIssuesListResponse, String> {
    validate_email(&params.email)?;
    validate_api_token(&params.api_token)?;
    let normalized_base_url = normalize_base_url(&params.base_url)?;

    let client = create_jira_client()?;
    let auth = build_basic_auth(&params.email, &params.api_token);
    let url = format!("{}/rest/api/3/search/jql", normalized_base_url);
    let body = json!({
        "jql": "assignee = currentUser() ORDER BY updated DESC",
        "maxResults": 20,
        "fields": ["key", "summary", "status", "assignee", "updated"]
    });

    let (status, data) = request_jira(&client, Method::POST, &url, &auth, Some(body))?;

    if status >= 400 {
        return Err(extract_jira_error(&data, status));
    }

    let issues = data
        .get("issues")
        .and_then(Value::as_array)
        .map(|items| {
            items
                .iter()
                .filter_map(|item| {
                    let key = item.get("key")?.as_str()?.to_string();
                    if key.trim().is_empty() {
                        return None;
                    }

                    let fields = item.get("fields").unwrap_or(&Value::Null);
                    let summary = fields
                        .get("summary")
                        .and_then(Value::as_str)
                        .unwrap_or("Untitled issue")
                        .to_string();
                    let status = fields
                        .get("status")
                        .and_then(|status| status.get("name"))
                        .and_then(Value::as_str)
                        .unwrap_or("Unknown")
                        .to_string();
                    let updated = fields
                        .get("updated")
                        .and_then(Value::as_str)
                        .map(ToOwned::to_owned);
                    let assignee = fields
                        .get("assignee")
                        .and_then(|assignee| assignee.get("displayName"))
                        .and_then(Value::as_str)
                        .map(ToOwned::to_owned);

                    Some(JiraIssue {
                        key,
                        summary,
                        status,
                        updated,
                        assignee,
                    })
                })
                .collect::<Vec<_>>()
        })
        .unwrap_or_default();

    Ok(JiraIssuesListResponse { issues })
}
